c++11/src/buffered_recorder.cpp

```diff
@@ -0,0 +1,166 @@
+                                   LightStepTracerOptions options,
+                                   std::unique_ptr<Transporter>&& transporter)
+    : logger_{logger},
+      options_{std::move(options)},
+      builder_{options_},
+      transporter_{std::move(transporter)} {
+  writer_ = std::thread(&BufferedRecorder::Write, this);
+}
+
+//------------------------------------------------------------------------------
+// Destructor
+//------------------------------------------------------------------------------
+BufferedRecorder::~BufferedRecorder() {
+  MakeWriterExit();
+  writer_.join();
+}
+
+//------------------------------------------------------------------------------
+// RecordSpan
+//------------------------------------------------------------------------------
+void BufferedRecorder::RecordSpan(collector::Span&& span) noexcept try {
+  std::lock_guard<std::mutex> lock_guard{write_mutex_};
+  if (builder_.num_pending_spans() >= options_.max_buffered_spans) {
+    dropped_spans_++;
+    return;
+  }
+  builder_.AddSpan(std::move(span));
+  if (builder_.num_pending_spans() >= options_.max_buffered_spans) {
+    write_cond_.notify_all();
+  }
+} catch (const std::exception& e) {
+  logger_.error("Failed to record span: {}", e.what());
+}
+
+//------------------------------------------------------------------------------
+// FlushWithTimeout
+//------------------------------------------------------------------------------
+bool BufferedRecorder::FlushWithTimeout(
+    std::chrono::system_clock::duration timeout) noexcept try {
+  // Note: there is no effort made to speed up the flush when
+  // requested, it simply waits for the regularly scheduled flush
+  // operations to clear out all the presently pending data.
+  std::unique_lock<std::mutex> lock{write_mutex_};
+
+  bool has_encoded = builder_.num_pending_spans() != 0;
+
+  if (!has_encoded && encoding_seqno_ == 1 + flushed_seqno_) {
+    return true;
+  }
+
+  size_t wait_seq = encoding_seqno_ - (has_encoded ? 0 : 1);
+
+  auto result = write_cond_.wait_for(lock, timeout, [this, wait_seq]() {
+    return write_exit_ || this->flushed_seqno_ >= wait_seq;
+  });
+  if (!result) {
+    return false;
+  }
+  return this->flushed_seqno_ >= wait_seq;
+} catch (const std::exception& e) {
+  logger_.error("Failed to flush recorder: {}", e.what());
+  return false;
+}
+
+//------------------------------------------------------------------------------
+// Write
+//------------------------------------------------------------------------------
+void BufferedRecorder::Write() noexcept try {
+  auto next = std::chrono::steady_clock::now() + options_.reporting_period;
+
+  while (WaitForNextWrite(next)) {
+    FlushOne();
+
+    auto end = std::chrono::steady_clock::now();
+    auto elapsed = end - next;
+
+    if (elapsed > options_.reporting_period) {
+      next = end;
+    } else {
+      next = end + options_.reporting_period - elapsed;
```

Comment:
I think the intention here is to try to catch up if the writer thread is taking longer than `options_.reporting_period` to run. For example, if it takes `2*options_.reporting_period`, then `elapsed == options_.reporting_period`, so the writer thread gets scheduled to run right away. Similarly, if it takes `1/2*options_.reporting_period`, `elapsed == 1/2*options_.reporting_period`, so it get scheduled to run next in half the time that it otherwise would.